zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,341 @@
+            }
+
+            if bits[2] {
+                tmp -= tmp.double();
+            }
+
+            if j > 0 {
+                // Inclusive range!
+                tmp *= (1..=(4 * j)).fold(jubjub::Fr::one(), |acc, _| acc.double());
+            }
+
+            m_i += tmp;
+        }
+
+        m_i
+    }
+
+    let mut result = jubjub::ExtendedPoint::identity();
+
+    // Split M into n segments of 3 * c bits, where c = 63, padding the last
+    // segment with zeros.
+    //
+    // https://zips.z.cash/protocol/protocol.pdf#concretepedersenhash
+    for (i, segment) in M.chunks(189).enumerate() {
+        result += I_i(domain, i as u32) * M_i(&segment);
+    }
+
+    result
+}
+
+/// Mixing Pedersen Hash Function
+///
+/// Used to compute ρ from a note commitment and its position in the note
+/// commitment tree.  It takes as input a Pedersen commitment P, and hashes it
+/// with another input x.
+///
+/// MixingPedersenHash(P, x) := P + [x]FindGroupHash^J^(r)(“Zcash_J_”, “”)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretemixinghash
+#[allow(non_snake_case)]
+pub fn mixing_pedersen_hash(P: jubjub::ExtendedPoint, x: jubjub::Fr) -> jubjub::ExtendedPoint {
+    const J: [u8; 8] = *b"Zcash_J_";
+
+    P + find_group_hash(J, b"") * x
+}
+
+/// Construct a 'windowed' Pedersen commitment by reusing a Pederson hash
+/// construction, and adding a randomized point on the Jubjub curve.
+///
+/// WindowedPedersenCommit_r (s) := \
+///   PedersenHashToPoint(“Zcash_PH”, s) + [r]FindGroupHash^J^(r)(“Zcash_PH”, “r”)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretewindowedcommit
+pub fn windowed_pedersen_commitment(r: jubjub::Fr, s: &BitVec<Lsb0, u8>) -> jubjub::ExtendedPoint {
+    const D: [u8; 8] = *b"Zcash_PH";
+
+    pedersen_hash_to_point(D, &s) + find_group_hash(D, b"r") * r
+}
+
+/// The randomness used in the Pedersen Hash for note commitment.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct CommitmentRandomness(jubjub::Fr);
+
+/// Note commitments for the output notes.
+#[derive(Clone, Copy, Deserialize, PartialEq, Serialize)]
+pub struct NoteCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
+
+impl fmt::Debug for NoteCommitment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("NoteCommitment")
+            .field("u", &hex::encode(self.0.get_u().to_bytes()))
+            .field("v", &hex::encode(self.0.get_v().to_bytes()))
+            .finish()
+    }
+}
+
+impl From<[u8; 32]> for NoteCommitment {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(jubjub::AffinePoint::from_bytes(bytes).unwrap())
+    }
+}
+
+impl From<jubjub::ExtendedPoint> for NoteCommitment {
+    fn from(extended_point: jubjub::ExtendedPoint) -> Self {
+        Self(jubjub::AffinePoint::from(extended_point))
+    }
+}
+
+impl From<NoteCommitment> for [u8; 32] {
+    fn from(cm: NoteCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
```

Comment:
Do we need both From impls on arrays *and* ZcashSerialize/ZcashDeserialize impls, or could we save code just by having one or the other? This is downstream of the internal representation issue.

Reply:
Probably just the latter 👍

Reply:
If either are equally good I have a slight preference for the former over the latter, because it just uses builtin traits rather than our custom ones.  In the transaction serialization code there's a bunch of examples of reading 32-byte data fields and then typing them (e.g., redjubjub keys) to see what it would look like.

Reply:
For `NoteCommitment`, which is never encoded on the wire (it's encrypted and the resulting `EncryptedCiphertext` is encoded), we definitely don't need `Zcash(De)Serialize`. `ValueCommitment` is encoded on the wire, but of course the meat of both impls is the same, so sure I'll keep `From` and remove `Zcash(De)Serialize`

Reply:
The u-coordinate of a note commitment is encoded (as `cmu` in an `OutputDescription`, and as a field element in commitment tree leaves in the Spend circuit). The commitment itself never is.